[8.x] Add throwWith to the HTTP client response

timacdonald opened the pull request:
Implementation updated: https://github.com/laravel/framework/pull/34558#issuecomment-699911645

Because:
- There is a repeated pattern where you want to perform a particular action if the response has failed, such as logging an error, **but still have the exception throw afterwards**.


Before...
```php
$response = $client->withHeaders($headers)->post($url, $payload);

if ($response->failed()) {
    Log::error('Twitter API failed posting Tweet', [
        'url' => $url,
        'payload' => $payload,
        'headers' => $headers,
        'response' => $response->body(),
    ]);

    $response->throw();
}

return $response->json();
```

After...
```php
return $client->withHeaders($headers)
    ->post($url, $payload)
    ->throwWith(fn ($response) =>
        Log::error('Twitter API failed posting Tweet', [
            'url' => $url,
            'payload' => $payload,
            'headers' => $headers,
            'response' => $response->body(),
        ])
    )->json();
```

This commit:
- Adds the `throwWith` method to the response.
- Updates the `throw` method to utilise the new `throwWith` method, just with an empty closure.

Notes:
- I'm not convinced this is the best name for this method.
- I'm also wondering if the method should be a `tap` method of some kind. `tapWhenThrowing` or something along those lines.

timacdonald committed:
Add throwWith to the HTTP client response

Because:
- There is a repeated pattern where you want to perform a particular
  action if the response has failed, such as logging the error.

```php
// before...

$response = $client->withHeaders($headers)->post($url, $payload);

if ($response->failed()) {
    Log::error('Twitter API failed posting Tweet', [
        'url' => $url,
        'payload' => $payload,
        'headers' => $headers,
        'response' => $response->body(),
    ]);

    $response->throw();
}

return $response->json();

// after...

return $client->withHeaders($headers)
    ->post($url, $payload)
    ->throwWith(fn ($response) =>
        Log::error('Twitter API failed posting Tweet', [
            'url' => $url,
            'payload' => $payload,
            'headers' => $headers,
            'response' => $response->body(),
        ])
    )->json();
```

This commit:
- Adds the `throwWith` method to the response.
- Updates the `throw` method to utilise the new `throwWith` method, just
  with an empty closure.

Notes:
- I'm not convinced this is the best name for this method.
- I'm also wondering if the method should be a `tap` method of some
  kind. `tapWhenThrowing` or something along those lines.

crynobone commented on src/Illuminate/Http/Client/Response.php:
```diff
@@ -204,12 +204,27 @@ public function toPsrResponse()
         if ($this->serverError() || $this->clientError()) {
+            $callback($this);
+
             throw new RequestException($this);
```
What if we do it as:

```php
throw tap(new RequestException($this), function ($exception) use ($callback) {
   $callback($this, $exception);
});
```

deleugpn commented on src/Illuminate/Http/Client/Response.php:
```diff
@@ -204,12 +204,27 @@ public function toPsrResponse()
+     *
+     * @throw \Illuminate\Http\Client\RequestException
+     */
+    public function throwWith($callback)
```
What about `beforeThrowing`?

m1guelpf replied:
Then it'd make sense to call `beforeThrowing()->throw()`, which IMO is too bervose

timacdonald replied:
agreed on that being to verbose. I was hoping to find something that could express "When throwing, do this thing and then throw" rather than two methods.

deleugpn replied:
`onError` also makes sense to me. `throwWith` doesn't make sense to me because it has an indication of which parameters would be given to `throw` (like the Cotnainer's `makeWith`). Though I still prefer `beforeThrowing` as it indicates what to do (callback) right before an exception is thrown. I guess `whenThrowing` could also work (like Query Builder `when()`).

timacdonald replied:
I don't disagree with your objections. As mentioned I didn't know if this was really a great method name.

Maybe there just isn't a nice "all-in-one" option, as `whenThrowing` also makes me wonder if I need to call `throw` or not.

Maybe it is just a `onFailure` callback (as others suggested on Twitter).

```php
return $client->withHeaders($headers)
    ->post($url, $payload)
    ->onFailure(fn ($response) =>
        Log::error('Twitter API failed posting Tweet', [
            'url' => $url,
            'payload' => $payload,
            'headers' => $headers,
            'response' => $response->body(),
        ])
    )->throw()->json();
```

Or maybe I just need to wrap up the more verbose middleware as @sebdesign pointed out in my own app 🤷‍♂️

I just liked how the failure and throw was on the other side of the `post` call, as it  _felt_ nicer to me, but that is just an opinion that might not be shared.

sebdesign replied:
Personally I prefer the `onError` callback instead of the `Middleware::mapResponse` stuff. It feels nicer and you get to work with the `Laravel\Http\Client\Response` instead of the Guzzle response.

GrahamCampbell commented on src/Illuminate/Http/Client/Response.php:
```diff
@@ -204,12 +204,27 @@ public function toPsrResponse()
      *
      * @return $this
-     *
-     * @throws \Illuminate\Http\Client\RequestException
```
this should be kept

timacdonald replied:
Thanks! Always forget if these should be applied to all calling places as well, or only the method throwing.

sebdesign commented:
I think the same can be achieved by using middleware.

Example:

```php
return $client->withHeaders($headers)
    ->middleware(Middleware::mapResponse(function ($response) {
        if ($response->getStatusCode() >= 400) {
            Log::error('Twitter API failed posting Tweet', [
                'url' => $url,
                'payload' => $payload,
                'headers' => $headers,
                'response' => (string) $response->getBody(),
            ]);
        }

        return $response;
    }))->post($url, $payload)->throw()->json();
```

timacdonald committed:
Implement `onError` instead of the combined `throwWith` method

Because:
- There wasn't a great name that really expressed what was happening
  under the hood, i.e. running the closure and throwing the exception

This commit:
- Reverts the `throw` method and implements a stand-alone `onError`
  method

timacdonald commented:
Although it isn't as nice as I had hoped, I've moved to an `onError` method that only deals with calling the callback and doesn't also throw the exception. Although we had some back and forth (https://github.com/laravel/framework/pull/34558#discussion_r495734812) I don't think we got an awesome option, so here we are.

**New implementation example:**

Before...
```php
$response = $client->withHeaders($headers)->post($url, $payload);

if ($response->failed()) {
    Log::error('Twitter API failed posting Tweet', [
        'url' => $url,
        'payload' => $payload,
        'headers' => $headers,
        'response' => $response->body(),
    ]);

    $response->throw();
}

return $response->json();
```

After...
```php
return $client->withHeaders($headers)
    ->post($url, $payload)
    ->onError(fn ($response) =>
        Log::error('Twitter API failed posting Tweet', [
            'url' => $url,
            'payload' => $payload,
            'headers' => $headers,
            'response' => $response->body(),
        ])
    )->throw()->json();
```

rodrigopedra commented on src/Illuminate/Http/Client/Response.php:
```diff
@@ -216,6 +216,21 @@ public function throw()
     }
 
+    /**
+     * Execute the callback and throw an exception if a server of client error occurred.
```
I guess the comment should be changed as an exception is not thrown anymore

deleugpn commented:
I think there's one last option for this that technically has to be split into 2 PRs:

PR 1 (8.x)
```
    public function throw()
    {
        $callback = func_get_arg(0);
            
        if ($this->serverError() || $this->clientError()) {
            if ($callback && is_callable($callback)) {
                $callback($this);
            }
            
            throw new RequestException($this);
        }

        return $this;
    }
```

PR 2 (9.x) 
```
    public function throw(?callable $callable = null)
    {
        if ($this->serverError() || $this->clientError()) {
            if ($callback) {
                $callback($this);
            }
            
            throw new RequestException($this);
        }

        return $this;
    }
```

**Usage: **

```
return $client->withHeaders($headers)
    ->post($url, $payload)
    ->throw(fn ($response) =>
        Log::error('Twitter API failed posting Tweet', [
            'url' => $url,
            'payload' => $payload,
            'headers' => $headers,
            'response' => $response->body(),
        ])
    )->json();
```

m1guelpf commented:
I like that, but I fear the name `throw` might lead people into believing you should throw the exception there instead of the callback running before the exception gets thrown.

deleugpn commented:
@m1guelpf I don't see that as a problem, but rather as a feature.

```
return $client->withHeaders($headers)
    ->post($url, $payload)
    ->throw(function ($response) {
        Log::error('Twitter API failed posting Tweet', [
            'url' => $url,
            'payload' => $payload,
            'headers' => $headers,
            'response' => $response->body(),
        ]);

        throw new MyCustomExceptionThatWillTakePrecedenceOverTheNativeException();
    })->json();
```

taylorotwell commented:
So I kept the `onError` method and also added support for passing a Closure to `throw` per @deleugpn suggestion. I kept both because I figured sometimes you may want to `onError` but not throw anything.

timacdonald commented:
Nice! Thanks for the collab everyone. Really dig where this landed with everyone’s input. Best of both worlds! ❤️
